tools/noescape/noescape.go

```diff
@@ -0,0 +1,435 @@
+// Copyright 2019 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package noescape allows recursive escape analysis for hot paths.
+//
+// The analysis tracks multiple types of escapes, in two categories. First,
+// 'hard' escapes are explicit allocations. Second, 'soft' escapes are
+// interface dispatches or dynamic function dispatches. The analysis is capable
+// of making assertions recursively: soft escapes cannot be analyzed in this
+// way, and therefore count as escapes for recursive purposes.
+//
+// The different types of escapes are as follows, with the category in
+// parentheses:
+//
+// 	interface: A call is made via an interface (soft).
```

Comment:
Why is this considered an escape? A call via an interface *may* cause an escape, but it itself is not an escape.

This may just be a documentation problem?

Reply:
See the paragraph above. I've added some more text here as well. Yes, it is a "soft" escape because we don't really know. The default +checkescape will only look for hard escapes, but you can also do +checkescape:all.